### DISCUSSION TITLE
Web assembly?

### DISCUSSION THREAD

**Discussion by Quafadas** [stdlib-js/stdlib#2779] [General]
Are any parts of the library compiled to web assembly? I was thinking in particular of the blas module… 

I could find no reference in the docs, so I’m just checking I’m not missing a trick here…

**Comment by kgryte** [comment]
@Quafadas We do support compiling to WebAssembly, but we are not currently shipping WebAssembly binaries. What is your use case and what BLAS functionality do you need?

**Comment by Quafadas** [comment]
It’s a curiosity now that anything else . I have an application which does some basic linear algebra (arrays only) and I was thinking about converting to wasm to see if there was a runtime benefit.

As things stand I think I have only level 1 functions in use. I wouldn’t want to absorb a lot of time.

Is that enough of a picture or should I provide more detail?

**Reply by Quafadas** [reply]
It does do a lot of array scaling so 20% gain would be noticeable

**Reply by kgryte** [reply]
Note that unless your application is entirely in WebAssembly OR you plan on allocating arrays directly on the wasm module heap, you'll lose any prospective perf gains due to shuttling memory back and forth between wasm and JS. Is your application publicly hosted on GitHub?

**Reply by Quafadas** [reply]
This is exactly the sorts of issues I want to explore. Hmmm. 

Currently, I don’t have an example I’m in a position to share. Which isn’t a great response - I’m sort of planning the experiment at the moment. I’ll go quiet for a couple of weeks - if it’s okay, I’m going to @ you when i have something concrete… bear with me!

**Reply by kgryte** [reply]
This is a bit dated at this point, but you may be interested in https://www.youtube.com/watch?v=iJL59lh4IJA&list=PLfMzBWSH11xa-iNnQG2555lgi4574nZOh&index=37.

**Reply by kgryte** [reply]
Unrelated to this discussion, I just did some performance comparisons of JS to other languages. For the benchmarks, I computed the sum using built-in language features (e.g., for Python, no NumPy; for C, no SIMD; etc). Here are the results, where the y-axis (Rate) is normalized relative to C and where languages are grouped along the x-axis according to array length. In short, using vanilla JS, V8, at least, can achieve near native performance once array sizes increase in size. All this to say, I'll be curious what you find during your WebAssembly experiments.
![language_comparison_benchmarks](https://github.com/user-attachments/assets/9999bd75-08d4-40b0-903f-98acba4381b3)

**Reply by Quafadas** [reply]
Gosh! That’s a cool picture. If that picture generalises, then I’m wasting my time!

My arrays are more or less Float64Array of size 10,000. I guess I’m wondering about the extent that sum could be aggressively optimised as it’s a common operation. Only way would be to measure my workflow I guess.

Could things like simd potential inn wasm make a difference? Although presumably, that’s more custom work?

To me, this seems to would imply that only parrallism , perhaps webworkers could help. And that is not to do with wasm or stdlib. Hmmmm.

**Reply by kgryte** [reply]
@Quafadas  Quick follow-up. Here is a benchmark with the BLAS level 1 routine `daxpy`. For the Wasm benchmark, computation is repeated on the same vectors allocated on the WebAssembly module heap. For the Wasm (copy) benchmark, data is copied into and out of the WebAssembly module heap (e.g., as might be the case when scripting or working with data coming from dynamic sources).

As can be seen from the chart, unless the FLOPS of the routine warrants the cost of marshalling data to and from Wasm memory, one can achieve equivalent, if not better, performance in just JavaScript.

...and this doesn't even address the increased complexity associated with memory fragmentation and unbounded growth which may occur when using Wasm modules.

Depends on your use case, but getting optimal Wasm performance can be a bit tricky to realize. I ran similar benchmarks on `dgemm` and there it was worth the perf penalty of marshalling data to and from the heap, as matmul is O(N^3) and copying data is O(N^2), where `N` corresponds to the number of elements along each dimension of a square matrix.

![daxpy_wasm_comparison_benchmarks](https://github.com/user-attachments/assets/9abe0e30-6263-457a-ae2e-d9875c157942)

**Reply by kgryte** [reply]
@Quafadas Quick update. We've started work on adding WebAssembly implementations of BLAS Level 1 routines. E.g., [`@stdlib/blas/base/daxpy-wasm`](https://github.com/stdlib-js/stdlib/tree/4728ffcffe8bc8d4ed9a1be332155ec7240b14e9/lib/node_modules/%40stdlib/blas/base/daxpy-wasm).

**Reply by kgryte** [reply]
Another update: we've continued working on WebAssembly APIs for various BLAS packages, which can be found in the BLAS namespace: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/blas/base. WebAssembly-specific package names end with `-wasm`.

**Reply by Quafadas** [reply]
@kgryte I did go away and do some homework. I watched the video, read your benchmark,  did some experiments, read other peoples experiments etc.

At the moment, the extra complexity of trying to get webassembly working, doesn't seem to be worth the uncertainty / perceived gain. I'm probably going to drop this for the time being. My plan will probably be to revisit it every now and again as webassembly evolves particulaly / when it becomes easier to get modules to interop? Currently, it seems a bit lower level than I want to be... I'll keep an eye out here though... 

And thank you so much, for the time and being willing to write and think about this 🙏 I do believe it could be beneficial in future!
